### DISCUSSION TITLE
Calculate tidal height using CSV file

### DISCUSSION THREAD

**Discussion by q1165600785** [pyTMD/pyTMD#352] [General]
Hello author, there are many options and settings for calculating tidal height. I want to use the position of the boundary point in CSV to calculate the monthly water level height, but I am not sure how CSV should be done correctly. Can you provide an example, author

**Comment by tsutterley** [comment]
hey @q1165600785, you're right the options have gotten a bit overwhelming. I _definitely_ need to work on the documentation as highlighted in #354. Particularly for people that are just getting started.

I'm trying to understand your use case. Are you wanting the _average_ monthly tide height, or the height at a given date during the month? Is this at a singular location or does it drift?

**Comment by q1165600785** [comment]
Hello, researcher，Hello, researcher, I uploaded two attachments, the latitude and longitude coordinates are the same, one is in MJD time format, the other is in python datetime format, my use case is a time series with an interval of one month per hour at each point of the ocean opening boundary, I tried to run it using pytmd, but it ended up with an error, and it ran very slowly, I found that utide python ran very fast, but I didn't get the correct result, I want to try using pytmd
[jb_mjd_test01.csv](https://github.com/user-attachments/files/17790122/jb_mjd_test01.csv)
[jb_test02_datatime.csv](https://github.com/user-attachments/files/17790124/jb_test02_datatime.csv)

**Comment by tsutterley** [comment]
Is there a specific model you are wanting to use? For ocean tides, pyTMD uses constants [from published models](https://pytmd.readthedocs.io/en/latest/getting_started/Getting-Started.html#model-database).

I think the basic syntax for your case should be:
```
compute_tidal_elevations.py --directory <path_to_tide_models> \
  --tide <tide_model_to_use> --crop --type drift \
  --format csv --header 1 --variables y x time --standard datetime \
  --infer-minor -VV jb_test02_datatime.csv
```

**Reply by q1165600785** [reply]
> 您想要使用特定模型吗？对于海洋潮汐，pyTMD 使用[已发布模型中的](https://pytmd.readthedocs.io/en/latest/getting_started/Getting-Started.html#model-database)常量。
> 
> 我认为你的情况的基本语法应该是：
> 
> ```
> compute_tidal_elevations.py --directory <path_to_tide_models> \
>   --tide <tide_model_to_use> --crop --type drift \
>   --format csv --header 1 --variables y x time --standard datetime \
>   --infer-minor -VV jb_test02_datatime.csv
> ```
Thank you for providing the demonstration reference. I will try running it using fes2022 or tpox10 to see what results will be obtained

**Comment by q1165600785** [comment]
@tsutterley 

Hello researcher, I used the eotides python (core is pytmd) package to calculate the tidal height, and I also used pyfvcom (which uses utide python internally) to calculate the tidal height. The tidal heights obtained by the two are completely different. Why is this? Are there any details that need to be paid attention to? I want to know why this is the case. Can you help me?

I used the same tide model TPXO10-atlas-v2-nc，But in pyfvcom I only use eight tidal components
constituents = ['K1', 'K2', 'M2', 'N2', 'O1', 'P1', 'Q1', 'S2']
Does this make such a big difference?


@robbibt  Developers, can etides add a parameter to the method for calculating tidal heights, so that you can select the tidal component you need, instead of including all tidal components in the calculation? Is my suggestion reasonable?

[obs_lat_lon_jb_test02_datatime20250414.csv](https://github.com/user-attachments/files/19970431/obs_lat_lon_jb_test02_datatime20250414.csv)
[obs_lat_lon_jb_test02_datatime20250414eotides.csv](https://github.com/user-attachments/files/19970434/obs_lat_lon_jb_test02_datatime20250414eotides.csv)

**Reply by robbibt** [reply]
@q1165600785 Are you able to reproduce this difference using `pyTMD` directly, with something like this?

```
compute_tidal_elevations.py --directory <path_to_tide_models> \
 --tide <tide_model_to_use> --crop --type drift \
 --format csv --header 1 --variables y x time --standard datetime \
 --infer-minor -VV jb_test02_datatime.csv
```

Would help isolate this as something caused by `eo-tides` or `pyTMD` itself.

A scatterplot of your two results would also be helpful.

I think being able to control what constituents are used would probably be something to be implemented on the `pyTMD` side... `eo-tides` is mainly just an EO-focused wrapper around `pyTMD` tools!

**Reply by q1165600785** [reply]
@robbibt   The API for calculating tidal heights in pytmd has become pyTMD.compute.tide_elevations.
This script ‘compute_tidal_elevations.py’ has been removed. I should use the above API, right?

**Reply by robbibt** [reply]
Something like this should be close to the `eo-tides` defaults:

```
from pyTMD.compute import tide_elevations

tide_elevations(
    x=..., 
    y=..., 
    delta_time=..., 
    DIRECTORY=..., 
    MODEL=..., 
    EPSG=4326, 
    TIME="datetime", 
    EXTRAPOLATE=True, 
    CUTOFF=np.inf, 
    METHOD="linear", 
    CROP=True,
) 
```

**Reply by tsutterley** [reply]
Right. I removed the file processing scripts as they were largely duplicative and `pyTMD` didn't support many file types (much better programs for that sort of thing!). @robbibt has the right function above. Would just need to read your csv (using e.g. `pandas`). I've been meaning to add a Recipes page to the documentation to show some specific examples.

**Comment by q1165600785** [comment]
@tsutterley   
Now the API for calculating tide height has become pyTMD.compute.tide_elevations. I uploaded the longitude and latitude, and I want to calculate the tide height every hour from March 1, 2024 to March 31, 2024. What should I do? I don't quite understand the required parameters.
[obs_lat_lon_only.csv](https://github.com/user-attachments/files/19973918/obs_lat_lon_only.csv)


`
import pyTMD
import  pandas as pd

def pro_tide_data():
    csv_file=r'C:\Users\DELL\Desktop\[公开]zyj20250425拷出\zyj20250425拷出\obs_lat_lon_only.csv'
    # csv_file=r'C:\Users\DELL\Desktop\[公开]zyj20250425拷出\zyj20250425拷出\obs_lat_lon_jb_test02_datatime20250414.csv'
    df=pd.read_csv(csv_file,)
    print(df)
    x=df['lon'].to_numpy()
    y=df['lat'].to_numpy()
    delta_time=df['lat'].to_numpy()
    DIRECTORY=r'E:\zyj20241216\TPXO10\tpxo10v2_netcdf'
    tide_h = pyTMD.compute.tide_elevations(x, y, delta_time,DIRECTORY=DIRECTORY, EPOCH=(2000,1,1,0,0,0),
        EPSG=3031, TYPE='drift')

if  __name__ =="__main__":
    pro_tide_data()
`


Traceback (most recent call last):
  File "E:\zyj20250320\pyTMD-main\pyTMD\io\model.py", line 247, in elevation
    self.from_dict(parameters['elevation'][m])
                   ~~~~~~~~~~~~~~~~~~~~~~~^^^
KeyError: None

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "E:\zyj20250320\pyTMD-main\cal_tidal_hegiht_TPOX10.py", line 18, in <module>
    pro_tide_data()
  File "E:\zyj20250320\pyTMD-main\cal_tidal_hegiht_TPOX10.py", line 14, in pro_tide_data
    tide_h = pyTMD.compute.tide_elevations(x, y, delta_time,DIRECTORY=DIRECTORY, EPOCH=(2000,1,1,0,0,0),
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\zyj20250320\pyTMD-main\pyTMD\compute.py", line 331, in tide_elevations
    model = pyTMD.io.model(DIRECTORY, compressed=GZIP).elevation(MODEL)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\zyj20250320\pyTMD-main\pyTMD\io\model.py", line 249, in elevation
    raise ValueError(f"Unlisted tide model {m}")
ValueError: Unlisted tide model None

Process finished with exit code 1

**Reply by tsutterley** [reply]
@robbibt had it right.

You need to describe your model type. Enter `MODEL='TPXO10-atlas-v2-nc'`

Use `EPSG=4326` instead of `EPSG=3031` (`3031` is polar stereographic south around Antarctica and not latitude/longitude).

Your csv file should have dates similar to your prior ones. Use `TIME='datetime'` and parse your csv file.

```
import pandas
import pyTMD
import numpy as np

df = pandas.read_csv('obs_lat_lon_jb_test02_datatime20250414.csv',parse_dates=['time'])
h = pyTMD.compute.tide_elevations(
    x=df.lon, 
    y=df.lat, 
    delta_time=df.time, 
    DIRECTORY=tide_model_directory, 
    MODEL="TPXO10-atlas-v2-nc", 
    EPSG=4326, 
    TIME="datetime", 
    EXTRAPOLATE=True, 
    CUTOFF=np.inf, 
    METHOD="linear", 
    CROP=True,
) 
```

If I read `obs_lat_lon_jb_test02_datatime20250414eotides.csv`, the above function from `pyTMD` gets identical values with your outputs from `eo-tides` (within the precision of the csv file).

If you want to modify a model (such as e.g. using less constituents), you can always create a `json` definition file to customize your model inputs [model_TPXO10-atlas-v2-reduced.json](https://github.com/user-attachments/files/19982242/model_TPXO10-atlas-v2-reduced.json).

I checked versus using your constrained list of constituents. Without knowing how you computed with `pyfvcom`, I cannot say why the results are different (except that the differences are quite significant).
